crates/store/src/db/models/queries/accounts/at_block.rs

```diff
@@ -0,0 +1,252 @@
+/// * `Ok(None)` - If account doesn't exist at that block
+/// * `Err(DatabaseError)` - If there's a database error
+pub(crate) fn select_account_header_at_block(
+    conn: &mut SqliteConnection,
+    account_id: AccountId,
+    block_num: BlockNumber,
+) -> Result<Option<AccountHeader>, DatabaseError> {
+    use schema::accounts;
+
+    let account_id_bytes = account_id.to_bytes();
+    let block_num_sql = block_num.to_raw_sql();
+
+    let account_data: Option<(AccountHeaderDataRaw, Option<Vec<u8>>)> = SelectDsl::select(
+        accounts::table
+            .filter(accounts::account_id.eq(&account_id_bytes))
+            .filter(accounts::block_num.le(block_num_sql))
+            .order(accounts::block_num.desc())
+            .limit(1),
+        (
+            (accounts::code_commitment, accounts::nonce, accounts::storage_header),
+            accounts::vault_root,
+        ),
+    )
+    .first(conn)
+    .optional()?;
+
+    let Some((
+        AccountHeaderDataRaw {
+            code_commitment: code_commitment_bytes,
+            nonce: nonce_raw,
+            storage_header: storage_header_blob,
+        },
+        vault_root_bytes,
+    )) = account_data
+    else {
+        return Ok(None);
+    };
+
+    let storage_commitment = match storage_header_blob {
+        Some(blob) => {
+            let storage = AccountStorage::read_from_bytes(&blob)?;
+            storage.to_commitment()
+        },
+        None => Word::default(),
+    };
+
+    let code_commitment = code_commitment_bytes
+        .map(|bytes| Word::read_from_bytes(&bytes))
+        .transpose()?
+        .unwrap_or(Word::default());
+
+    let nonce = nonce_raw.map_or(Felt::ZERO, raw_sql_to_nonce);
+
+    let vault_root = vault_root_bytes
+        .map(|bytes| Word::read_from_bytes(&bytes))
+        .transpose()?
+        .unwrap_or(Word::default());
+
+    Ok(Some(AccountHeader::new(
+        account_id,
+        nonce,
+        vault_root,
+        storage_commitment,
+        code_commitment,
+    )))
+}
+
+// ACCOUNT CODE
+// ================================================================================================
+
+/// Queries the account code for a specific account at a specific block number.
+///
+/// Returns `None` if:
+/// - The account doesn't exist at that block
+/// - The account has no code (private account or account without code commitment)
+///
+/// # Arguments
+///
+/// * `conn` - Database connection
+/// * `account_id` - The account ID to query
+/// * `block_num` - The block number at which to query the account code
+///
+/// # Returns
+///
+/// * `Ok(Some(Vec<u8>))` - The account code bytes if found
+/// * `Ok(None)` - If account doesn't exist or has no code
+/// * `Err(DatabaseError)` - If there's a database error
+pub(crate) fn select_account_code_at_block(
+    conn: &mut SqliteConnection,
+    account_id: AccountId,
+    block_num: BlockNumber,
+) -> Result<Option<Vec<u8>>, DatabaseError> {
```

Comment:
Related to the above comment: I don't think we need this method. Since we already know code commitment for a given account header, we can look up account code by commitment. So, what we need instead is something like `select_account_code_by_commitment()` (unless we already have it).